[GFC] Populate eligibility function for layout with some initial logic

sammygill opened the pull request:
#### c22965919816960450cc8290c6aac3da01b89aca
<pre>
[GFC] Populate eligibility function for layout with some initial logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=300145">https://bugs.webkit.org/show_bug.cgi?id=300145</a>
<a href="https://rdar.apple.com/161924265">rdar://161924265</a>

Reviewed by Alan Baradlay.

Currently, our canUseForGridLayout eligibility function returns false
whether or not the feature flag is enabled. Now, since we have an
initial implementation that can support some very restricted content, we
should populate this function to allow it. Over time, we will ease out
the conditions here as we claim more and more content to be supported
inside of GFC.

The nature of the content that is currently allowed is essentially fixed
sized grids and items in which the grid is explicitly sized, and the
items are all explicitly placed. There are some additional pieces that
restrict content (e.g. only allows start alignment currently), but this
is the idea at a high level.

Certain AvoidanceReasons are very generic just for this initial set of
conditions but will likely become more fine-grained as we continue to
complete GFC. For example, we return GridHasUnsupportedGridTemplateColumns
in all cases where we run into grids that have any sort of column
styling outside of fixed sizes. Over time, we will likely remove this and
add additional ones that are more descriptive of the condition that
caused us to bail out.

Canonical link: <a href="https://commits.webkit.org/301063@main">https://commits.webkit.org/301063@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd9b1bdd9f72fbba9586654e12aa3beff2f93ab6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44268 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35002 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131435 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76539 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126473 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52839 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94787 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62856 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35854 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111425 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75359 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34790 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29580 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74914 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105609 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29811 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134103 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39266 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103266 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51856 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107647 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103044 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26288 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48405 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26669 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48378 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51319 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57115 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156266 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50718 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/156266 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54073 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52407 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/adf3df7dea0543487401c07c56e94b8b72c37817. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51784

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd9b1bdd9f72fbba9586654e12aa3beff2f93ab6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51784

sammygill committed:
[GFC] Populate eligibility function for layout with some initial logic
https://bugs.webkit.org/show_bug.cgi?id=300145
rdar://161924265

Reviewed by Alan Baradlay.

Currently, our canUseForGridLayout eligibility function returns false
whether or not the feature flag is enabled. Now, since we have an
initial implementation that can support some very restricted content, we
should populate this function to allow it. Over time, we will ease out
the conditions here as we claim more and more content to be supported
inside of GFC.

The nature of the content that is currently allowed is essentially fixed
sized grids and items in which the grid is explicitly sized, and the
items are all explicitly placed. There are some additional pieces that
restrict content (e.g. only allows start alignment currently), but this
is the idea at a high level.

Certain AvoidanceReasons are very generic just for this initial set of
conditions but will likely become more fine-grained as we continue to
complete GFC. For example, we return GridHasUnsupportedGridTemplateColumns
in all cases where we run into grids that have any sort of column
styling outside of fixed sizes. Over time, we will likely remove this and
add additional ones that are more descriptive of the condition that
caused us to bail out.

Canonical link: https://commits.webkit.org/301063@main
